File: graph/segment_tree.cpp
/* poj 3468 "A Simple Problem with Integers" */
#include <iostream>
using namespace std;
typedef long long ll;
const char nl = '\n';
const int N = 1e5+10;

// segment tree allowing interval modifying
ll sum[N<<2], add[N<<2]; // 4x space
void push_up(int rt) { // update upwards
    sum[rt] = sum[rt<<1] + sum[rt<<1|1]; // rt*2, rt*2+1
}
void push_down(int rt, int m) { // update children of rt  m: number of elements belonging to node rt
    if (add[rt]) {
        add[rt<<1] += add[rt];
        add[rt<<1|1] += add[rt];
        sum[rt<<1] += (m-(m>>1))*add[rt];
        sum[rt<<1|1] += (m>>1)*add[rt];
        add[rt] = 0; // cancel tagging
    }
}
// left subtree
#define ltree l, mid, rt<<1
// right subtree
#define rtree mid+1, r, rt<<1|1
void build(int l, int r, int rt) { // build the segment tree using a full bst
    add[rt] = 0;
    if (l == r) { // leaf, assign value
        cin >> sum[rt];
        return;
    }
    int mid = l + ((r-l)>>1);
    build(ltree); build(rtree);
    push_up(rt); // update interval sum upwards
}
void update(int a, int b, ll c, int l, int r, int rt) {
    if (a <= l && b >= r) {
        sum[rt] += (r-l+1)*c;
        add[rt] += c; // tagging
        return;
    }
    push_down(rt, r-l+1); // updating downwards
    int mid = l + ((r-l)>>1);
    if (a <= mid) update(a, b, c, ltree); // require updating in the left subtree
    if (b > mid)  update(a, b, c, rtree); // require updating in the right subtree
    push_up(rt);
}
ll query(int a, int b, int l, int r, int rt) {
    if (a <= l && b >= r) return sum[rt]; // lazy is satisfied, return directly
    push_down(rt, r-l+1); // cannot lazy anymore
    int mid = l + ((r-l)>>1);
    ll ans = 0;
    if (a <= mid) ans += query(a, b, ltree);
    if (b > mid)  ans += query(a, b, rtree);
    return ans;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    int n, m;
    cin >> n >> m;
    build(1, n, 1);
    while (m--) {
        char s;
        cin >> s;
        int a, b; ll c;
        if (s == 'C') {
            cin >> a >> b >> c;
            update(a, b, c, 1, n, 1);
        } else {
            cin >> a >> b;
            cout << query(a, b, 1, n, 1) << nl;
        }
    }
    return 0;
}

File: pyds/segment_tree.py
class Node:  # struct of segment tree node
    def __init__(self, v, l, r):
        self.v = v
        self.l, self.r = l, r
        self.add = 0 # for lazy-tagging
        self.child = [None, None]

    def __str__(self):
        return f"({self.v})[{self.l},{self.r}] " + (str(self.child[0]) if self.child[0] else '') + (str(self.child[1]) if self.child[1] else '')

def push_up(node):
    node.v = node.child[0].v + node.child[1].v

def push_down(node):
    if node.add:
        add = node.add
        mid = (node.l+node.r)//2
        node.child[0].add += add
        node.child[1].add += add
        node.child[0].v += add*(mid-node.l+1)
        node.child[1].v += add*(node.r-mid)
        node.add = 0 # cancel tagging

def build(l, r): # build a segment tree, return its root
    root = Node(0, l, r)
    if l == r: return root
    mid = (l+r)//2
    root.child = [build(l, mid), build(mid+1, r)]
    return root

def update(node, l, r, a):
    assert node.l <= l and r <= node.r
    if l == node.l and node.r == r:
        node.v += a*(r-l+1)
        node.add += a # lazy-tagging
        return
    push_down(node)
    mid = (node.l+node.r)//2
    if l <= mid: update(node.child[0], l, mid, a)
    if r > mid: update(node.child[1], mid+1, r, a)
    push_up(node)

def query(node, l, r):
    if l <= node.l and node.r <= r: return node.v
    push_down(node)
    mid = (node.l+node.r)//2
    res = 0
    if l <= mid: res += query(node.child[0], l, r)
    if r > mid: res += query(node.child[1], l, r)
    return res


if __name__ == '__main__':
    # test
    tree = build(1, 8)
    print(tree)
    update(tree, 1, 8, 5)
    print(query(tree, 1, 8)) # 40
    print(tree)
    update(tree, 1, 4, 3)
    print(query(tree, 1, 8)) # 52
    print(tree)
    print(query(tree, 1, 5)) # 37
